lib/OnyxCache.js

```diff
@@ -0,0 +1,183 @@
+import _ from 'underscore';
+
+const isDefined = _.negate(_.isUndefined);
+
+/**
+ * In memory cache providing data by reference
+ * Encapsulates Onyx cache related functionality
+ */
+class OnyxCache {
+    constructor() {
+        /**
+         * @private
+         * Cache of all the storage keys available in persistent storage
+         * @type {Set<string>}
+         */
+        this.storageKeys = new Set();
+
+        /**
+         * @private
+         * A map of cached values
+         * @type {Record<string, *>}
+         */
+        this.storageMap = {};
+
+        /**
+         * @private
+         * Captured pending tasks for already running storage methods
+         * @type {Record<string, Promise>}
+         */
+        this.pendingPromises = {};
+
+        // bind all methods to prevent problems with `this`
+        _.bindAll(
+            this,
+            'getAllKeys', 'getValue', 'hasCacheForKey', 'update', 'remove', 'merge', 'resolveTask',
+        );
+    }
+
+    /**
+     * Get all the storage keys
+     * Provide a promise returning fallback that will be executed when
+     * keys aren't available in cache
+     * @param {function(): Promise<string[]>} fallback - a callback to retrieve keys from storage
+     * @returns {Promise<string[]>}
+     */
+    getAllKeys(fallback) {
```

Comment:
NAB, I still think passing in `fallback` here and for `get()` is unnecessary. We are trying to avoid the cache having a dependency on `AsyncStorage`, but I'm not convinced there's any reason to do so.

Reply:
Thanks for your thoughts. Let's go ahead and make this change.

Reply:
Updated how get and getAllKeys work and it should be clear how data retrieving works and when cache is involved